lib/azuremetadata/azuremetadata.py

```diff
@@ -15,25 +23,73 @@ def get_all(self):
+            "http://169.254.169.254/metadata/instance?api-version={}".format(quote(self._api_version))
+        )
 
     def get_attested_data(self):
-        # FIXME this is a stub while the CLI UX is being figured out
-
-        if self._api_version == '2019-08-15':
-            with open('fixtures/attested-data-v2019-08-15.json', 'r') as json_file:
-                return json.load(json_file)
-        elif self._api_version == '2018-10-01':
-            with open('fixtures/attested-data-v2018-10-01.json', 'r') as json_file:
-                return json.load(json_file)
-        else:
-            raise Exception("Unknown API version fixture exception (attested data)")
+        return self._make_request(
+            "http://169.254.169.254/metadata/attested/document?api-version={}".format(quote(self._api_version))
+        )
+
+    def get_disk_tag(self, device=None):
+        if not device:
+            device = self._find_root_device()
+
+        if not device:
+            return ''
+
+        try:
+            with open(device, 'rb') as fh:
+                fh.seek(65536)
+                return str(uuid.UUID(bytes_le=fh.read(16)))
+        except OSError as e:
+            print("An error occurred when reading disk tag:", file=sys.stderr)
+            print(str(e), file=sys.stderr)
+            return ''
+
+    @staticmethod
+    def _find_root_device():
+        hex_device_id = os.stat("/").st_dev
+        root_device_id = str(os.major(hex_device_id)) + ":" + str(os.minor(hex_device_id))
+
+        devices = glob.glob("/sys/block/*")
+        for device_path in devices:
+            device = os.path.basename(device_path)
+            partitions = glob.glob("/sys/block/{}/{}*".format(device, device))
+
+            for partition in partitions:
+                with open("{}/dev".format(partition), "r") as fh:
+                    device_id = fh.read().strip()
+
+                if device_id == root_device_id:
+                    return "/dev/{}".format(device)
+
+        return None
+
+    @staticmethod
+    def _make_request(url):
+        tries = 0
+        last_error = None
+        while tries < 5:
+            try:
+                req = urllib.request.Request(url, headers={'Metadata': 'true'})
+                response = urllib.request.urlopen(req, timeout=1)
+
+                data = response.read()
+                if isinstance(data, bytes):
+                    data = data.decode('utf-8')
+                return json.loads(data)
+            except urllib.error.HTTPError as e:
+                print("An error occurred when fetching metadata:", file=sys.stderr)
+                print(str(e), file=sys.stderr)
+                print(e.read(), file=sys.stderr)
+                return {}
+            except OSError as e:
+                tries += 1
+                last_error = e
+                sleep(1)
+
+        print("An error occurred when fetching metadata:", file=sys.stderr)
+        print(str(last_error), file=sys.stderr)
```

Comment:
use of str, better use format, and in this case it's not required at all because done by print

Reply:
Fair enough, explicit `str` isn't needed, removed.